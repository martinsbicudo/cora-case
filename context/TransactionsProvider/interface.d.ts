import { Children, Dispatch, SetStateAction } from 'react'

import transactions from '@/data/transactions.json'

export interface TransactionsProviderProps {
  children: Children
}

export type TransactionsType = typeof transactions | []

export type FiltersType = {
  search: string
  filter: 'all' | 'credit' | 'debit' | 'future'
}

export interface TransactionsContextType {
  transactions: TransactionsType
  setTransactions: Dispatch<SetStateAction<TransactionsType>>
  filters: FiltersType
  setFilters: Dispatch<SetStateAction<FiltersType>>
}
