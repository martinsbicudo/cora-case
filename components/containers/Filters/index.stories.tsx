import { StoryObj, Meta } from '@storybook/react'

import Filters from '.'

export default {
  title: 'Containers / Filters',
  component: Filters,
} as Meta

export const Default: StoryObj = {}
