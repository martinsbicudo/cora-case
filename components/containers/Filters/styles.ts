import { up } from 'styled-breakpoints'
import styled from 'styled-components'

export const Filters = styled.section`
  padding: 25px 0;

  & > *:last-child {
    margin-top: 20px;
  }

  ${up('md')} {
    display: flex;
    align-items: center;
    flex-direction: row-reverse;
    padding: 40px 0;

    & > *:last-child {
      flex: 0;
      margin-right: 26px;
      margin-top: 0;
    }
  }
`
