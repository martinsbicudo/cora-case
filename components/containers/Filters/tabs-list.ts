import { ItemType } from '@Commons/Tabs/interface'

const tabsList = [
  {
    key: 'all',
    label: 'Tudo',
  },
  {
    key: 'credit',
    label: 'Entrada',
  },
  {
    key: 'debit',
    label: 'Sa√≠da',
  },
  {
    key: 'future',
    label: 'Futuro',
  },
] as ItemType[]

export default tabsList
