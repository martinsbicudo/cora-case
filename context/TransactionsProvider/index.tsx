import { useState, useMemo, createContext } from 'react'

import {
  TransactionsProviderProps,
  TransactionsContextType,
  TransactionsType,
  FiltersType,
} from './interface'

export const TransactionsContext = createContext<TransactionsContextType>(
  {} as TransactionsContextType
)

const TransactionsProvider = ({
  children,
}: TransactionsProviderProps): JSX.Element => {
  const [initialTransactions, setTransactions] = useState<TransactionsType>([])
  const [filters, setFilters] = useState<FiltersType>({
    search: '',
    filter: 'all',
  })

  const checkActorIncludesSearch = (actor: string, search: string) =>
    actor.toLowerCase().includes(search.toLowerCase())

  const transactions = useMemo(() => {
    const { search, filter } = filters

    let filteredTransactions = [...initialTransactions]
    filteredTransactions = filteredTransactions.map((transaction) => {
      let filteredItems = [...transaction.items]

      switch (filter) {
        case 'credit':
          filteredItems = filteredItems.filter(
            ({ actor, entry }) =>
              entry === 'CREDIT' && checkActorIncludesSearch(actor, search)
          )
          break
        case 'debit':
          filteredItems = filteredItems.filter(
            ({ actor, entry }) =>
              entry === 'DEBIT' && checkActorIncludesSearch(actor, search)
          )
          break
        case 'future':
          filteredItems = filteredItems.filter(
            ({ actor, scheduled }) =>
              scheduled && checkActorIncludesSearch(actor, search)
          )
          break
        default:
          filteredItems = filteredItems.filter(({ actor }) =>
            checkActorIncludesSearch(actor, search)
          )
          break
      }

      return {
        ...transaction,
        items: filteredItems,
      }
    })

    return filteredTransactions.filter(({ items }) => Boolean(items?.length))
  }, [filters, initialTransactions])

  const value = useMemo(
    () => ({ transactions, setTransactions, filters, setFilters }),
    [transactions, setTransactions, filters, setFilters]
  )

  return (
    <TransactionsContext.Provider value={value}>
      {children}
    </TransactionsContext.Provider>
  )
}

export default TransactionsProvider
