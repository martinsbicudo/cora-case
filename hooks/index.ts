export { default as useTransactions } from './use-transactions'
