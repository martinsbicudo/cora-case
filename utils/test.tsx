import type { ReactElement, ReactNode } from 'react'

import { ThemeProvider } from 'styled-components'

import theme from '@/styles/theme'
import { render, RenderOptions } from '@testing-library/react'

const Providers = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export { customRender as render }
