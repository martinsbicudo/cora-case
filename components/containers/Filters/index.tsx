import { ChangeEvent } from 'react'

import { FiltersType } from '@/context/TransactionsProvider/interface'
import { useTransactions } from '@/hooks'
import { Tabs, SearchInput } from '@Commons'

import * as Styled from './styles'
import tabsList from './tabs-list'

const Filters = () => {
  const { setFilters } = useTransactions()

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    const search = e.target.value

    setFilters((currentFilters) => ({
      search,
      filter: currentFilters.filter,
    }))
  }

  const handleTabsChange = (filter: FiltersType['filter']) => {
    setFilters((currentFilters) => ({
      search: currentFilters.search,
      filter,
    }))
  }

  return (
    <Styled.Filters>
      <SearchInput onChange={handleInputChange} />
      <Tabs items={tabsList} onChange={handleTabsChange} />
    </Styled.Filters>
  )
}

export default Filters
