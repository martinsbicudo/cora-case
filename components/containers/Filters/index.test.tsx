import { FiltersType } from '@/context/TransactionsProvider/interface'
import { useTransactions } from '@/hooks'
import { render, screen, fireEvent } from '@/test/helpers'

import Filters from '.'

jest.mock('@/hooks')

describe('Containers / Filters', () => {
  const spySetFilters = jest.fn()
  const initialFilters = {
    search: '',
    filter: 'all',
  } as FiltersType
  const renderComponent = () => render(<Filters />)

  beforeAll(() => {
    ;(useTransactions as jest.Mock).mockImplementation(() => ({
      setFilters: (action: (filters: FiltersType) => FiltersType) =>
        spySetFilters(action(initialFilters)),
    }))
  })

  it('should render component', () => {
    const { container } = renderComponent()
    expect(container).toMatchSnapshot()
  })

  it('should call setFilters on dispatch input change', () => {
    const value = 'test'
    renderComponent()

    const input = screen.getByRole('textbox')
    fireEvent.change(input, { target: { value } })
    expect(spySetFilters).toBeCalledWith({
      search: value,
      filter: 'all',
    })
    spySetFilters.mockClear()
  })

  it('should call setFilters on dispatch tab click', () => {
    renderComponent()

    const [, button] = screen.getAllByRole('button')
    fireEvent.click(button)

    expect(spySetFilters).toBeCalledWith({
      search: '',
      filter: 'credit',
    })
  })
})
