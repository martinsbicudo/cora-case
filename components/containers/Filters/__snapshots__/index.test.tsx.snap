// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers / Filters should render component 1`] = `
<div>
  <section
    class="sc-jrcTuL dDzSDY"
  >
    <fieldset
      class="sc-eDvSVe dAMRCn"
    >
      <div
        class="sc-jSUZER Sxxek"
      >
        <i>
          icon-mock
        </i>
      </div>
      <input
        class="sc-gKPRtg ccSwvI"
        id="search"
        name="search"
        placeholder="Pesquisar"
        type="text"
      />
    </fieldset>
    <ul
      class="sc-iBYQkv bEoIEo"
    >
      <li
        class="sc-ftTHYK uEAGe"
      >
        <button
          class="sc-pyfCe MYluQ"
          type="button"
        >
          Tudo
        </button>
      </li>
      <li
        class="sc-ftTHYK uEAGe"
      >
        <button
          class="sc-pyfCe OCrcI"
          type="button"
        >
          Entrada
        </button>
      </li>
      <li
        class="sc-ftTHYK uEAGe"
      >
        <button
          class="sc-pyfCe OCrcI"
          type="button"
        >
          Sa√≠da
        </button>
      </li>
      <li
        class="sc-ftTHYK uEAGe"
      >
        <button
          class="sc-pyfCe OCrcI"
          type="button"
        >
          Futuro
        </button>
      </li>
    </ul>
  </section>
</div>
`;
